finetune/datasets/generated/dyck_loader.py

```diff
@@ -0,0 +1,164 @@
+                # Add the close character to the to be used close character list
+                unused_close_characters.append(dyck_char_pair[1])
+            else:
+                # Choose a random unused close character to use now.
+                close_index = random.randint(0, len(unused_close_characters) - 1)
+                close_char = unused_close_characters.pop(close_index)
+                dyck_word += close_char
+                close_count += 1
+
+        return dyck_word
+
+    def __init__(
+        self,
+        dyck_character_pairs: typing.List[
+            typing.Tuple[str, str]
+        ] = DYCK_CHARACTER_PAIRS,
+        min_length_pairs: int = 2,
+        max_length_pairs: int = 20,
+        min_length_answer: int = 1,
+        max_length_answer: int = 3,
+        random_seed: typing.Optional[int] = None,
+        samples: int = 100,
+    ):
+        """Loads prompt/response data from generated dyck words.
+
+        Args:
+            dyck_character_pairs (typing.List[ typing.Tuple[str, str] ], optional): Pairs of open/close characters to use. Defaults to DYCK_CHARACTER_PAIRS.
+            min_length_pairs (int, optional): Minimum number of pairs to create the dyck word from. Defaults to 2.
+            max_length_pairs (int, optional): Maximum number of pairs to create the dyck word from. Defaults to 20.
+            min_length_answer (int, optional): Minimum length of the reference answer. Defaults to 1.
+            max_length_answer (int, optional): Maximum length of the reference answer. Defaults to 3.
+            random_seed (typing.Optional[int], optional): Seed to use for all random operations if set.
+            samples (int, optional): Number of samples to generate. Defaults to 100.
+        """
+        self.buffer: typing.List[typing.Tuple[str, str]] = []
+
+        if random_seed is not None:
+            random.seed(random_seed)
+
+        for _ in range(samples):
+            pair_count = random.randint(min_length_pairs, max_length_pairs)
+            complete_dyck = self._generate_dyck(dyck_character_pairs, pair_count)
+
+            reference_length = random.randint(min_length_answer, max_length_answer)
+            # Ensure we don't make the reference longer than the number of pairs.
+            if reference_length > pair_count:
+                reference_length = pair_count
+
+            # Randomly remove close characters into the reference answer up to length.
+            # Get index of all the close characters.
+            close_chars = {pair[1] for pair in DYCK_CHARACTER_PAIRS}
+            close_char_indexes = []
+            for index, char in enumerate(complete_dyck):
+                if char in close_chars:
+                    close_char_indexes.append(index)
+
+            chosen_indexes = random.sample(close_char_indexes, reference_length)
+            # Sort the chosen indexes so the reference answer has them in the right order.
+            chosen_indexes.sort()
```

Comment:
Note the sorting doesn't actually matter as we enumerate in order below.

However it is also not clear to my that this is ever deterministic.

Both ( < ) > and ( < < ) are valid dyck words so given ( < it is not clear what the reference should be.

Reply:
Resolved. For a Dyck language with multiple delimiters a stack approach is used.